Added example for ADMM

ThomasDeb opened the pull request:
Example with a custom solver for x-minimization step

ThomasDeb committed:
Minor documentation changes

ThomasDeb committed:
Added example for ADMM

ThomasDeb committed:
Rename operators in ADMM example

ThomasDeb committed:
Big fixes in DouglasRachford and ProximalPoint

ThomasDeb committed:
Minor updates and performance improvement

ThomasDeb committed:
Added tests for all PDS algos

joanrue reviewed (commented):
These tests are really great and well thought. 
My only comment is that there is a lot of repeated code that could be encapsulated into a base class. I've drafted this idea for some of the classes: 

```
class MixinPDS(conftest.SolverT):

    @pytest.fixture
    def spec(self, klass, init_kwargs, fit_kwargs) -> tuple[pyct.SolverC, dict, dict]:
        return klass, init_kwargs, fit_kwargs

    @pytest.fixture
    def N(self) -> int:
        return 7

    @pytest.fixture(params=["1d", "nd"])
    def x0(self, N, request):
        # multiple initial points
        return {"1d": np.full((N,), 3.0),
                "nd": np.full((2, N), 15.0)}[request.param]

    @pytest.fixture
    def klass(self) -> pyct.SolverC:
        return NotImplementedError

    @pytest.fixture
    def init_kwargs(self) -> dict:
        return NotImplementedError

    @pytest.fixture
    def fit_kwargs(self, x0) -> dict:
        # Overriden only for ADMM
        return dict(
            x0=x0,
        )


class TestCV(MixinPDS):

    @pytest.fixture
    def klass(self) -> pyct.SolverC:
        return pycos.CV

    @pytest.fixture
    def init_kwargs(self, N: int) -> dict:

        K = PSDConvolution(N=N)
        K.lipschitz()
        return dict(
                f=pycf.SquaredL2Norm(dim=N), g=pycf.shift_loss(pycf.SquaredL2Norm(dim=N), 1), h=pycf.L1Norm(dim=N), K=K
            )

class TestPD3O(MixinPDS):
    @pytest.fixture
    def klass(self) -> pyct.SolverC:
        return pycos.PD3O

    @pytest.fixture
    def init_kwargs(self, N) -> dict:
        from pycsou_tests.operator.examples.test_posdefop import PSDConvolution
        K = PSDConvolution(N=N)
        K.lipschitz()
        return dict(
            f=pycf.SquaredL2Norm(dim=N), g=pycf.shift_loss(pycf.SquaredL2Norm(dim=N), 1), h=pycf.L1Norm(dim=N), K=K
        )

class TestCP(MixinPDS):
    @pytest.fixture
    def klass(self) -> pyct.SolverC:
        return pycos.CP

    @pytest.fixture
    def init_kwargs(self, N) -> dict:
        from pycsou_tests.operator.examples.test_posdefop import PSDConvolution
        K = PSDConvolution(N=N)
        K.lipschitz()
        return dict(
            g=pycf.shift_loss(pycf.SquaredL2Norm(dim=N), 1), h=pycf.L1Norm(dim=N), K=K
        )


class TestLV(MixinPDS):
    @pytest.fixture
    def klass(self) -> pyct.SolverC:
        return pycos.LV

    @pytest.fixture
    def init_kwargs(self, N) -> dict:
        from pycsou_tests.operator.examples.test_posdefop import PSDConvolution
        K = PSDConvolution(N=N)
        K.lipschitz()
        return dict(  # Inner-loop NLCG (with prox)
            f=pycf.SquaredL2Norm(dim=N), h=pycf.shift_loss(pycf.L1Norm(dim=N), 1), K=K
        )

ThomasDeb commented:
Ah yeah you're totally right, thanks for the suggestion! I'll implement that ASAP (that's how it's supposed to go, right?)

ThomasDeb committed:
Treat f=None case for ADMM

ThomasDeb committed:
Implemented Joan's suggestion to have a superclass for all PDS tests

ThomasDeb committed:
Update of generate_funcs :
- reduce number of specs by avoiding unnecessary permutation of elements that are summed together
- add diff_lipschitz computation, which is necessary for testing PDS algos

ThomasDeb committed:
Updated tests for PD3O and ADMM to conform with new solver mathematical test API. All other PDS algos still need to be updated

joanrue committed:
[pds] update with QuadraticFunc refactoring

joanrue committed:
[pds][BugFix] tau for tuning_strategy==2 must be smaller than 2/beta

joanrue committed:
[pds][BugFix] remove input reshaping and monkey patch PD3O corner case

joanrue committed:
[pds][tests] add remaining solvers & fixtures (base and tuning strategy)

ThomasDeb committed:
Cleaner definition of dimension of problems in PDS tests

ThomasDeb committed:
Update QuadraticFunc __init__ parameters in ADMM

ThomasDeb committed:
Update ADMM tests to to test NLCG scenario

SepandKashani committed:
[op][func] Add alias-import of QuadraticFunc.

SepandKashani committed:
[util] Add import_module() to ease dynamic module loads.

SepandKashani committed:
[op][lipschitz] Interface change for Lin[Op,Func]().

Summary:
* LinOp.lipschitz(recompute, algo, **kwargs) is replaced by LinOp.lipschitz(tight, **kwargs).
* `Intent` of .lipschitz() is modified: returns a Lipschitz constant for LinOps, not necessarily the optimal value.

SepandKashani committed:
[transpose] Add dedicated test suite.

All .T-related tests in operator/conftest.py now tested in arithmetic/test_transposerule.py

SepandKashani committed:
[test][op][lipschitz] Drop hidden attribute checks.

Users are not supposed to query ._[diff_]lipschitz.
Moreover these quantities are not handled well by arithmetic.
Since users must always go through .[diff_]lipschitz() to obtain a (diff-)Lipschitz constant, this test should be dropped in favor of test_math_[diff_]lipschitz().

SepandKashani committed:
[op][BugFix] Move transpose() logic to arithmetic.

SepandKashani committed:
[arithmetic][asloss] Propagate only for AddRule/ScaleRule.

2023.03.21: Pycsou Dev Meeting Consensus
.asloss() applied to arg[scale,shift]() is difficult to interpret and error-prone.

It is thus disabled.

SepandKashani committed:
[test][norm][BugFix] Drop domain-agnostic SquaredL2 instantiation.

Thought it had been fixed already? Apparently not.

SepandKashani committed:
[test][op] Add precCM tests for [diff_]lipschitz(), trace().

SepandKashani committed:
[op] Enforce precision on [diff_]lipschitz(), trace().

matthieumeo committed:
checkpoint commit

SepandKashani committed:
[op][stencil] Lift precision-specific constraint (with warning).

Stencils only work with input dtypes which match the kernel's.
We purposely do NOT want to force a copy for performance reasons.
However, it is common for novice users to not take care of precision issues, thus stumble when using Stencil().

Decision: allow using alternative dtype-inputs, but warn user if sub-optimal casting required.

ThomasDeb committed:
[pds][tests] Add convergence tests for dual problem

SepandKashani committed:
[from_sciop] Move to pycsou.operator.interop

from_sciop() is currently located in LinOp.from_sciop().
Problem: more "wrapper" functions will be provided in the future to interface with PyTorch, Jax, etc.

In order to group all wrapper functionality in the same namespace, from_sciop() is moved to pycsou.operator.interop.

SepandKashani committed:
[test][from_sciop] Drop rotten disable_test entries.

SepandKashani committed:
[test][trace] Precision bugfix

matthieumeo committed:
[util][array_module] Decorator for wrapping non-Dask functions to make them weakly compatible with Dask (Dask input -> compute (delayed) -> processing (delayed) -> Dask ouput (from delayed).

Does not seem to work with methods for some unclear reason.

matthieumeo committed:
[interop][pytorch] First working prototype of a Pycsou Map wrapper for PyTorch.

SepandKashani committed:
[SWE] Finufft is an optional dependency (again).

If a user does not have it installed, then linop.nufft still can be imported, but not run.

SepandKashani committed:
[inspect] import_module() optionally suppresses import failures.

SepandKashani committed:
[nufft] Conditionally load library.

SepandKashani committed:
[test][svd] Precision bugFix

SepandKashani committed:
[test][BugFix] Allow marking flaky tests inline.

SepandKashani committed:
[test][lipschitz] Precision bugfix

SepandKashani committed:
[stencil][BugFix] trace() on GPU must return a scalar.

SepandKashani committed:
[arithmetic][chainrule][BugFix] Lipschitz constant of LinFuncs is always computed (cheaply).

SepandKashani committed:
[SWE] Update working notes.

SepandKashani committed:
[doc] Add pycsou.operator.interop.

SepandKashani committed:
[interop] Add from_source().

matthieumeo committed:
[interop][pytorch] First working prototype of a Pycsou Map wrapper for PyTorch with autodiff and adjoint features.

matthieumeo committed:
[interop][pytorch][BugFix] Fix bug in arguments of _delayed_if_dask decorators.

SepandKashani committed:
[from_source][BugFix] _expr() defaults to that provided by cls.

SepandKashani committed:
Supersede direct use of `types.MethodType` by `interop.source.from_source`.

SepandKashani committed:
[SWE] Update working notes.

matthieumeo committed:
[interop] Moved PyTorch wrapper under operator module with other wrappers.

SepandKashani committed:
[from_source] Allow embedding of extra information.

SepandKashani committed:
[from_source] auto-vectorize support

SepandKashani committed:
[doc][BugFix] to_sciop() moved to interop.sciop.py

SepandKashani committed:
[util][operator] Add _dask_zip().

Low-level function to create parallel/serial dask graphs when needed.

This function is an enhanced variant of the _scan() method from nufft.py.
nufft.py should be later refactored to use _dask_zip().

SepandKashani committed:
[util][operator] Enhance vectorize().

Now accepts a `method` and `codim` parameter to tailor vectorization strategy.

3 modes are supported: scan, scan_dask, and parallel.

See the updated docstring for more details.

SepandKashani committed:
[from_source] Allow parameterization of vectorize().

Lets users specify vectorization method, including adding support for DASK inputs.

SepandKashani committed:
[doc][lipschitz] Mention why xp/gpu dropped for certain backends.

SepandKashani committed:
[from_source][API] codim-spec unnecessary.

vectorize() codim parameter differs per arithmetic method.
It is now inferred based on `shape` and the arithmetic method it is applied to.
`vkwargs` is replaced with `vmethod`, where different vectorization strategies may be specified per method.

SepandKashani committed:
[from_source][API] Allow FP-precision enforcement.

Pycsou arithmetic methods must abide to the runtime context manager.
The `enforce_precision` option allows one to auto-decorate user-provided function as needed to enable this.

SepandKashani committed:
[test][op] Ensure [diff_]lipschitz() return correct types.

SepandKashani committed:
[DiagonalOp][BugFix] lipschitz() must return a CPU float.

SepandKashani committed:
[block][BugFix] Do not propagate arithmetic attributes.

Reason: ABC classes define suitable defaults; they will be used instead of hard-coded np.infs from `__init__` which may break .lipschitz() methods from some ABCs.

Ex: if a block operator is inferred to be unitary, then by the old system its lipschitz constant would be inferred to be np.inf (because np.inf defaults from block.__init__ were used.)

The new code keeps arithmetic attributes located in `op`.

SepandKashani committed:
[block][BugFix] [diff_]lipschitz() should use cheap defaults when applicable.

lipschitz() applied on Lin[Op,Func]s would always compute the Lipschitz constant.
Now it defaults to a cheap upper-bound, with fallback on LinOp.lipschitz() if `tight` kwargs is provided.

SepandKashani committed:
[arithmetic] Safeguard: don't do in-place updates of [diff_]lipschitz().

matthieumeo committed:
[doc][images] Updated Pycsou UML diagram.

matthieumeo committed:
[notebooks] Added BIG presentation + relocated all notebooks in notebooks/.

matthieumeo committed:
[notebooks] Deleted relocated notebook.

SepandKashani committed:
[norm] Use vectorize where applicable.

SepandKashani committed:
[block,nufft] Rename _wrap_if_dask() -> _parallelize()

The goal of this function is to optionally parallelize execution of apply/grad/etc when fed with NUMPY inputs.
It assumes the function knows how to handle DASK arrays.

SepandKashani committed:
[from_source][BugFix] Auto-fill missing vmethod-spec when dict() provided.

SepandKashani committed:
[_dask_zip][BugFix] Functions may have side-effects if parallel=False.

SepandKashani committed:
[HomothetyOp][BugFix] dagger() is not an arithmetic method.

SepandKashani committed:
[DiagonalOp][BugFix] dagger() is not an arithmetic method.

SepandKashani committed:
[SWE] Add JAX as optional dependency.

SepandKashani committed:
[from_source][BugFix] Precision can be enforced on (diff_)lipschitz(), [svd,eig]vals(), trace().

SepandKashani committed:
[Resilience] Use pycd.NDArrayInfo to access array modules.

SepandKashani committed:
[interop.source] Use OO architecture to ease maintenance.

SepandKashani committed:
[from_source][doc] Enhance docstrings.

SepandKashani committed:
[from_source] _FromSource() does not take default arguments.

Similar to _COOBlock(), it is the job of the public interface from_source() to specify everything to the low-level _FromSource() class.

SepandKashani committed:
[interop] Add JAX module.

SepandKashani committed:
[doc] Add interop.jax.

SepandKashani committed:
[jax] Add JAX interop.

SepandKashani committed:
[jax] test WIP

ThomasDeb committed:
[pds][doc] Update path of QuadraticFunc in documentation

ThomasDeb committed:
Merge branch 'v2-dev' into v2-ADMM
